tb_plugin/tensorboard_plugin_torch_profiler/profiler/module_parser.py

```diff
@@ -300,14 +315,30 @@ def parse_kernels(kernel_list):
 
             return kernel_list_groupby_name_op
 
-        tid2list = {}
+        # For OperatorNode and ProfilerStepNode:
+        #   Use time interval containing relationship to build father-child correlation,
+        #   which is consistent with autograd profiler.
+        # For RuntimeNode:
+        #   Use external_id to build correlation with its father OperatorNode or ProfilerStepNode.
+        #   Because in the case when RuntimeNode has duration 0 and starts at same time as a OperatorNode,
+        #   just use interval containing relationship can't tell it is child or brother of the OperatorNode.
+        tid2list = {}  # value is a list of OperatorNode and ProfilerStepNode. Do not include RuntimeNode
+        tid2rt_list = {}  # value is a list of RuntimeNode with external_id=0. They will be attached to root nodes.
         corrid_to_device = {}  # value is a list of DeviceNode
         corrid_to_runtime = {}  # value is a RuntimeNode
+        externalid_to_runtime = {}  # value is a list of RuntimeNode
         for event in events:
-            parse_event(event, corrid_to_device, corrid_to_runtime, tid2list)
-        for tid, host_node_list in tid2list.items():
-            host_node_list.sort(key=lambda x: (x.start_time, x.end_time))
-            root_node = self._build_tree(host_node_list)
+            parse_event(event, corrid_to_device, corrid_to_runtime, externalid_to_runtime, tid2list, tid2rt_list)
+        # associate CUDA Runtimes with CPU events
+        for _, op_list in tid2list.items():
+            for op in op_list:
+                if op.external_id in externalid_to_runtime:
+                    op.runtimes.extend(externalid_to_runtime[op.external_id])
```

Comment:
Do you want to add a check that if any runtime has external id that doesn't correspond to any op?